src/scanner.c

```diff
@@ -0,0 +1,172 @@
+#define adv lexer->advance(lexer, false)
+#define eof lexer->eof(lexer)
+
+enum TokenType {
+    RAW_STRING_BEGIN,
+    RAW_STRING_CONTENT,
+    RAW_STRING_END,
+    ERROR_SENTINEL,
+};
+
+typedef struct {
+    uint8_t level;
+    bool emitted_content;
+} Scanner;
+
+static uint8_t consume_chars(TSLexer *lexer, char c) {
+    uint8_t count = 0;
+    while (lexer->lookahead == c && !eof) {
+        adv;
+        ++count;
+    }
+    return count;
+}
+
+static uint32_t consume_until(TSLexer *lexer, char c) {
+    uint32_t count = 0;
+    while (lexer->lookahead != c && !eof) {
+        adv;
+        count++;
+    }
+    return count;
+}
+
+
+void *tree_sitter_nu_external_scanner_create(void) {
+    Scanner *scanner = ts_malloc(sizeof(Scanner));
+    scanner->level = 0;
+    scanner->emitted_content = false;
+    return scanner;
+}
+
+
+void tree_sitter_nu_external_scanner_destroy(void *payload) {
+    ts_free(payload);
+}
+
+
+unsigned tree_sitter_nu_external_scanner_serialize(
+    void *payload,
+    char *buffer
+) {
+    Scanner *s = (Scanner *) payload;
+    buffer[0] = s->level;
+    buffer[1] = s->emitted_content;
+    return 2;
+}
+
+
+void tree_sitter_nu_external_scanner_deserialize(
+    void *payload,
+    const char *buffer,
+    unsigned length
+) {
+    Scanner *s = (Scanner *) payload;
+    s->level = 0;
+    s->emitted_content = false;
+    if (length == 2) {
+        s->level = buffer[0];
+        s->emitted_content = buffer[1];
+    }
+}
+
+static void skip_whitespace(TSLexer *lexer) {
+    while ((lexer->lookahead ==  ' ' || lexer->lookahead == '\t'
+            || lexer->lookahead == '\n') && !eof) {
+        skip;
+    }
+}
+
+static bool scan_raw_string_begin(TSLexer *lexer, Scanner *s) {
+    lexer->log(lexer, "BEGIN\n");
+    // scan for r#' r##' or more #
+    skip_whitespace(lexer);
+
+    if (lexer->lookahead == 'r') {
+        lexer->log(lexer, "Detected 'r'.\n");
+        adv;
+        uint8_t level = consume_chars(lexer, '#');
+        lexer->log(lexer, "num #: %i\n", level);
+        if (lexer->lookahead == '\'') {
+            adv;
+            lexer->log(lexer, "Detected level: %i\n", level);
+            s->level = level;
+            return true;
+        }
+    }
+    return false;
+}
+
+static bool scan_raw_string_content(TSLexer *lexer, Scanner *s) {
+    uint32_t len = 0;
+    lexer->log(lexer, "CONTENT\n");
+    while (!eof) {
+        uint32_t num = consume_until(lexer, '\'');
+        lexer->log(lexer, "Set mark\n");
+        lexer->mark_end(lexer);
+        len += num;
+        adv;
+        uint8_t level = consume_chars(lexer, '#');
+        lexer->log(lexer, "Consumed [%i] #\n", level);
+        if (level == s->level && len != 0) {
+            lexer->log(lexer, "Detected end\n" );
+            return true;
+        } else if (level == s->level && len == 0) {
+            return false;
+        }
+    }
+
+    return false;
+}
+
+bool scan_raw_string_end(TSLexer *lexer, Scanner *s) {
```

Comment:
I can pass all your test cases with the following code:
```c
static bool scan_raw_string_end(TSLexer *lexer, Scanner *s) {
    lexer->log(lexer, "END\n");
    adv;
    for (unsigned i = 0; i < s->level; i++) {
        adv;
    }
    s->level = 0;
    return true;
}
```

<img width="366" alt="image" src="https://github.com/user-attachments/assets/706ed699-74b6-4114-9681-18fee6513bde">

Better to put the `raw-strings.nu` in `expr` ?

Reply:
Let me try to explain my understanding of the logic here:

1. in `scan_raw_string_content` we already make sure the ending level is correct,
   - if it's not, for example, at the 2nd  `'` in `r##' hello'# '##`, it won't stop the while loop, it will advance further until the last `'` is reached, then mark_end there.
2. then in `scan_raw_string_end`, we simply advance `s->level` + 1 (for the `'`) times.